File: justmc-jdk/src/main/java/justmc/GameValue.java
package justmc;

import justmc.annotation.Inline;

@Inline
public enum GameValue {
    CURRENT,
    DEFAULT,
    DEFAULT_ENTITY,
    KILLER,
    DAMAGER,
    VICTIM,
    SHOOTER,
    PROJECTILE,
    LAST_ENTITY;

    public static native GameValue ofPlayer(Player player);
    public static native GameValue ofEntity(Entity entity);
    public static native Primitive get(String id);

    public native Primitive getValue(String id);
}

File: justmc-code/src/main/kotlin/me/unidok/justcode/value/GameValue.kt
package me.unidok.justcode.value

import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive

data class GameValue(
    val id: String,
    val selector: Selector? = null
) : Value {
    enum class Selector {
        CURRENT,
        DEFAULT,
        DEFAULT_ENTITY,
        KILLER,
        DAMAGER,
        VICTIM,
        SHOOTER,
        PROJECTILE,
        LAST_ENTITY
    }

    override fun serialize(): JsonObject = JsonObject(mapOf(
        "type" to JsonPrimitive("game_value"),
        "game_value" to JsonPrimitive(id),
        "selection" to JsonPrimitive(selector?.let {
            JsonObject(mapOf("type" to JsonPrimitive(it.name.lowercase())))
        }.toString())
    ))

    override fun toString(): String {
        return if (selector == null) "%val($id)"
        else "%val($id,$selector)"
    }
}